README.md
## Chef Server Populator

Creates orgs, clients, and admin users and installs provided public keys. Simplifies managing and
recreating Chef Server nodes. Provides backup and restore recipes for
complete Chef Server recovery.

### New Chef 12 Support
Chef 12 is supported in version 1.0 and above. If you need Chef 11
support, please pin your environment to version 0.4.0.

### Usage

**When bootstrapping with the chef-server cookbook and chef-solo:**

* Download and unpack chef-server, chef-server-ingredient, packagecloud, and chef-server-populator cookbooks
* Upload public keys to be used by users, org-validator, and clients (optionally)
* Create json for organization, user, and (optionally) client(s)
* Run chef-solo

See the `default[:chef_server_populator][:solo_org]` and
`default[:chef_server_populator][:solo_org_user]` attribute hashes in
`attributes/default.rb` for the required attribute structure.

**When converging with chef-client:**

* Create data bag to hold data bag items with user, org, and client information
* Create data bag items with user, org, and client information
* Set data bag related attributes

Applicable attributes:

* `node[:chef_server_populator][:databag]` - name of the data bag

Structure of the data bag item:

User:
```json
{
  "id": "user_name",
  "chef_server": {
    "full_name": "User Name",
    "email": "name@domain.tld",
    "client_key": "public key contents",
    "type": [
      "user"
    ],
    "orgs": {
      "organization": {
        "enabled": true,
        "admin": true
      }
    }
  }
}
```
Note: While users can belong to multiple organizations, and the above
hash structure allows you to define multiple associations, the
chef-server-populator currently only supports the first organization
that is defined in the data bag.

Client:
```json
{
  "id": "client_name",
  "chef_server": {
    "client_key": "public key contents",
    "type": [
      "client"
    ],
    "orgs": [ "organization" ]
   }
}
```
Note: If no organization is specified for a client, it will be added
to the default organization. The client `enabled` and `admin` settings
can be set at the top level of the `chef_server` hash or in and `orgs`
hash as in the User example.

Org:
```json
{
  "id": "org_name",
  "chef_server": {
    "full_name": "Organization Name",
    "client_key": "public key contents",
    "type": [
      "org"
    ],
    "enabled": true
  }
}
```
Note: Creating the org will create a client called `<org>-validator` which uses the public key specified when
creating the org.
In addition, there is currently a bug in Chef server 12.1 which means only the first word in
the full name will be used, as the option is not parsed correctly

**Restoring from a backup:**

* Set path to restore file with node[:chef_server_populator][:restore][:file]
* The restore recipe is run if a restore file is set
* The restore file can be remote or local

**When enabling backups:**

* Include chef-server-populator::restore recipe
* Set backup cron interval with node[:chef_server_populator][:schedule]
* Optionally set a remote storage location with node[:chef_server_populator][:backup][:remote][:connection]
* Backups include both a pg_dump of the entire chef database and a tarball of the Chef data directory

## Public Key Format

The format of the public key specified with the json object needs to be a single line string with new lines
represented with the \n character

You can use one of the below commands to convert your public key file into the correct string format (credit
to the certificates cookbook for these)
```
cat <filename> | sed s/$/\\\\n/ | tr -d '\n'
-OR-
/usr/bin/env ruby -e 'p ARGF.read' <filename>
-OR-
perl -pe 's!(\x0d)?\x0a!\\n!g' <filename>
```
If you need to obtain the public key string for your private key first, then run the following on the .pem
file containing the private key
```
openssl rsa -in <path_to_keyfile>.pem -pubout
```

## Extras

Need to use the IP address of the node for a bit, or another name  instead of
having `node[:fqdn]`?

* `node[:chef_server_populator][:servername_override]`

Keep chef server configured via chef client:

* `node[:chef_server_populator][:chef_server]`

If the hash is non-empty, it will write the chef-server `dna.json` and trigger a
`reconfigure` when ever the attributes are updated.

## Known Issues

* As mentioned above, user and client data bag items currently only
  support the first organization provided. Multi-org support is
  forthcoming.

## Examples

Take a look in the `examples` directory for basic bootstrap templates that will
build a new erchef server, using existing keys and client, and
register itself, or restore an existing chef server from a backup.

## Info
* Repository: https://github.com/hw-cookbooks/chef-server-populator
* IRC: Freenode @ #heavywater


CHANGELOG.md
## Unreleased changes

## v2.0.0
* Support chef-server 4.x versions
* Make chef-server constraint tighter to prevent breakage
* Remove resource notifications to resources outside our recipes
* Refactor configuration setup to support hash style attribute and auto-convert to string

## v1.2.2
* Adds test for user key
* Fixes org recipe user key in 12.1.x

## v1.2.0
* Adds support for both 12.0.x and 12.1.x versions, client key
  commands got a new flag which broke things.
* Prevents randomly generated passwords from starting with a '-' which
  will register as a flag and cause an error.
* Fixes backup recipe when using remote backups
* Fixes backup recipe when running as cron
* Updates and extends the integration tests to cover backups

## v1.1.4
* Fixes all users created as admins
* Updates user creation to require explicit enabled setting
* Updates client admin to default to false

## v1.1.2
* Adds myriad unit tests.
* Fixes issue #23 where users were not assigned to orgs.
* Fixes issue #22 where client recipe failed if no chef_server hash set.

## v1.1.0
* Fixes and loosens idempotency tests to account for Chef Server
version differences
* Updates Client recipe to create orgs, then users, then clients
* Fixes non-existent org attribute in solo recipe
* Fixes missing user keys in test suite
* Moves common org/user/client creation specs to the spec helper
* Refactors Backup/Restore to work with Chef 12. Uses Miasma rather
than Fog.

## v1.0.2
* Org recipe only included for solo run, since client run expects data
bag items.

## v1.0.0
* Updates to support Chef 12
* Removes support for Chef 11
* Adds support for organization creation in solo and client contexts
* Updates backup/restore recipes for new psql path and new table &
field names. (Not fully tested)
* Replaces many knife and psql commands with native chef-server-clt
management commands

## v0.4.0
* Allow for creation of clients, users, or both
* Store backup configuration in separate JSON file
* Provide proper retries to account for temporary server unavailability
* Include full server restart on restore from backup
* Provide 'latest' backup files along with named files
* Convert backup script from template to cookbook file
* Make service restarts more consistent

## v0.3.2
* Add client creation retries to stabilize initial bootstrap
* Updates to example bootstrap script
* Add support for backup/restore (thanks @luckymike!)

## v0.3.0
* Include chef-server dependency
* Update configuration overrides for chef-server
* Use `:endpoint` attribute for custom hostname/ip

## v0.2.0
* Provide distinct solo vs. client recipes
* Client recipe configures dna.json and uses ctl for reconfigure


recipes/client.rb
include_recipe 'chef-server-populator::configurator'

knife_cmd = "#{node[:chef_server_populator][:knife_exec]}"
knife_opts = '-c /etc/opscode/pivotal.rb'

ssl_port = %w(chef-server configuration nginx ssl_port).inject(node) do |memo, key|
  memo[key] || break
end
ssl_port = ":#{ssl_port}" if ssl_port

pg_cmd = "/opt/chef-server/embedded/bin/psql -d opscode_chef"

if(node[:chef_server_populator][:databag])
  begin
    items = data_bag(node[:chef_server_populator][:databag]).map do |bag_item|
      item = data_bag_item(node[:chef_server_populator][:databag], bag_item).fetch('chef_server', {})
      if item.empty?
        Chef::Log.info("No chef-server data for #{bag_item['id']}")
      end
      item.merge('client' => data_bag_item(node[:chef_server_populator][:databag], bag_item)['id'],
                 'pub_key' => item['client_key'],
                 'enabled' => item['enabled'],
                 'admin' => item.fetch('admin', false),
                 'password' => item.fetch('password', SecureRandom.urlsafe_base64(23).gsub(/^\-*/,'')),
                 'orgs' => item.fetch('orgs', {}))
    end
    orgs = items.select { |item| item.fetch('type', []).include?('org') }
    users = items.select { |item| item.fetch('type', []).include?('user') }
    clients = items.select { |item| item.fetch('type', []).include?('client') }

    # Org Setup
    orgs.each do |item|
      if(item['enabled']==true)
        item['full_name'] = item.fetch('full_name', item['client'].capitalize)
        execute "create org: #{item['client']}" do
          item.merge('full_name' => item.fetch('full_name', item['client'].capitalize))
          command "chef-server-ctl org-create #{item['client']} #{item['full_name']}"
          not_if "chef-server-ctl org-list | grep '^#{item['client']}$'"
          if item['client'] == node[:chef_server_populator][:default_org]
            notifies :reconfigure, 'chef_server_ingredient[chef-server-core]', :immediately
          end
        end
        if(item['pub_key'])
          key_file = "#{Chef::Config[:file_cache_path]}/#{item['client']}.pub"
          file key_file do
            backup false
            content item['pub_key']
            mode '0400'
          end
        end
        execute "add org validator key: #{item['client']}" do
          if (node['chef-server'][:version].to_f >= 12.1 || node['chef-server'][:version].to_f == 0.0)
          command "chef-server-ctl add-client-key #{item['client']} #{item['client']}-validator --public-key-path #{key_file} --key-name populator"
          else
            command "chef-server-ctl add-client-key #{item['client']} #{item['client']}-validator #{key_file} --key-name populator"
          end
          not_if "chef-server-ctl list-client-keys #{item['client']} #{item['client']}-validator | grep 'name: populator$'"
        end
        execute "remove org default validator key: #{item['client']}" do
          command "chef-server-ctl delete-client-key #{item['client']} #{item['client']}-validator default"
          only_if "chef-server-ctl list-client-keys #{item['client']} #{item['client']}-validator | grep 'name: default$'"
        end
      else
        Chef::Log.info("#{item['client']} is not enabled, skipping.")
      end
    end
    # User Setup
    users.each do |item|
      org, options = item['orgs'].first
      item['org'] = org
      if(options)
        if(options.has_key?('enabled'))
          item['enabled'] = options['enabled']
        end
        if(options.has_key?('admin'))
          item['admin'] = options['admin']
        end
      end
      if(item['enabled'] == false)
        execute "remove user: #{item['client']} from #{item['org']}" do
          command "chef-server-ctl org-user-remove #{item['org']} #{item['client']}"
        end
        execute "delete user: #{item['client']}" do
          command "chef-server-ctl user-delete #{item['client']}"
          only_if "chef-server-list user-list | tr -d ' ' | grep '^#{item['client']}$'"
        end
      elsif(item['enabled'] == true)
        if(item['pub_key'])
          unless item['pub_key'].to_s.empty?
            key_file = "#{Chef::Config[:file_cache_path]}/#{item['client']}.pub"
            file key_file do
              backup false
              content item['pub_key']
              mode '0400'
            end
            item['full_name'] = item.fetch('full_name', item['client'].capitalize)
            first_name = item['full_name'].split(' ').first.capitalize
            last_name = item['full_name'].split(' ').last.capitalize
            email = item.fetch('email', "#{item['client']}@example.com")
            execute "create user: #{item['client']}" do
              command "chef-server-ctl user-create #{item['client']} #{first_name} #{last_name} #{email} #{item['password']} > /dev/null 2>&1"
              not_if "chef-server-ctl user-list | grep '^#{item['client']}$'"
            end
            execute "set user key: #{item['client']}" do
              if (node['chef-server'][:version].to_f >= 12.1 || node['chef-server'][:version].to_f == 0.0)
                command "chef-server-ctl add-user-key #{item['client']} --public-key-path #{key_file} --key-name populator"
              else
                command "chef-server-ctl add-user-key #{item['client']} #{key_file} --key-name populator"
              end
              not_if "chef-server-ctl list-user-keys #{item['client']} | grep 'name: populator$'"
            end
            execute "delete default user key: #{item['client']}" do
              command "chef-server-ctl delete-user-key #{item['client']} default"
              only_if "chef-server-ctl list-user-keys #{item['client']} | grep 'name: default$'"
            end
            execute "set user org: #{item['client']}" do
              command "chef-server-ctl org-user-add #{item['org']} #{item['client']} #{'--admin' if item['admin']}"
            end
          end
        end
      end
    end
    # Client Setup
    clients.each do |item|
      org, options = item['orgs'].first
      if(org)
        knife_url = "-s https://127.0.0.1/organizations/#{org}"
      else
        knife_url = "-s https://127.0.0.1"
      end
      if(options)
        if(options.has_key?('enabled'))
          item[:enabled] = options[:enabled]
        end
        if(options.has_key?('admin'))
          item[:admin] = options[:admin]
        end
      end
      if(item['enabled'] == false)
        execute "delete client: #{item['client']}" do
          command "#{knife_cmd} client delete #{item['client']} -d #{knife_opts} #{knife_url}"
          only_if "#{knife_cmd} client list #{knife_opts} #{knife_url} | tr -d ' ' | grep '^#{item['client']}$'"
          retries 10
        end
      else
        if(item['pub_key'])
          key_file = "#{Chef::Config[:file_cache_path]}/#{item['client']}.pub"
          file key_file do
            backup false
            content item['pub_key']
            mode '0400'
          end
        end
        execute "create client: #{item['client']}" do
          command "#{knife_cmd} client create #{item['client']}#{' --admin' if item['admin']} -d #{knife_url} #{knife_opts}"
          not_if "#{knife_cmd} client list #{knife_url} #{knife_opts} | tr -d ' ' | grep '^#{item['client']}$'"
          retries 10
        end
        if(item['pub_key'])
          execute "set client key: #{item['client']}" do
            if (node['chef-server'][:version].to_f >= 12.1 || node['chef-server'][:version].to_f == 0.0)
              command "chef-server-ctl add-client-key #{org || node[:chef_server_populator][:default_org]} #{item['client']} --public-key-path #{key_file} --key-name populator"
            else
              command "chef-server-ctl add-client-key #{org || node[:chef_server_populator][:default_org]} #{item['client']} #{key_file} --key-name populator"
            end
              not_if "chef-server-ctl list-client-keys #{org || node[:chef_server_populator][:default_org]} #{item['client']} | grep 'name: populator$'"
          end
          execute "delete default client key: #{item['client']}" do
            command "chef-server-ctl delete-client-key #{org || node[:chef_server_populator][:default_org]} #{item['client']} default"
            only_if "chef-server-ctl list-client-keys #{org || node[:chef_server_populator][:default_org]} #{item['client']} | grep 'name: default$'"
          end
        end
      end
    end
  rescue Net::HTTPServerException
    Chef::Log.warn 'Chef server populator failed to locate population data bag'
  end
end
